src/dfx/src/main.rs

```diff
@@ -170,7 +174,29 @@ fn print_error_and_diagnosis(err: Error, error_diagnosis: Diagnosis) {
 }
 
 fn main() {
-    let cli_opts = CliOpts::parse();
+    let matches = {
+        let installed_extensions = ExtensionManager::command();
+        let mut app =
+            CliOpts::command_for_update().subcommands(installed_extensions.get_subcommands());
+        if !installed_extensions.get_subcommands().count() > 0 {
+            sort_clap_commands(&mut app);
+        }
+        app.get_matches()
+    };
+
+    let mut args = std::env::args_os().collect::<Vec<_>>();
+
+    if let Ok(em) = ExtensionManager::new(dfx_version(), false) {
+        // safe to unwrap because clap will display help if no subcommand is provided
+        // which happens when "app.get_matches()" is called above
+        let subcmd = matches.subcommand().unwrap().0;
+        if !DfxCommand::has_subcommand(subcmd) && em.is_extension_installed(subcmd) {
+            args.insert(1, "extension".into());
+            args.insert(2, "run".into());
```

Comment:
The benefit of this approach is that we don't need to rewrite anything below (e.g. setting up logging works). The extension shorthand notation will get converted into `DfxCommand::Extension`, which would not be possible otherwise